include/seqan3/io/alignment_file/format_bam.hpp

```diff
@@ -806,27 +851,41 @@ class alignment_file_output_format<format_bam> : alignment_file_output_format<fo
             // ---------------------------------------------------------------------
             // Writing the Record
             // ---------------------------------------------------------------------
+            int32_t ref_length{};
 
-            // compute possible distance from alignment end to sequence end
-            // which indicates soft clipping at the end.
-            // This should be replaced by a free count_gaps function for
-            // aligned sequences which is more efficient if possible.
-            auto off_end{std::ranges::distance(seq) - offset};
+            // if alignment is non-empty, replace cigar_vector.
+            // else, compute the ref_length from given cigar_vector which is needed to fill field `bin`.
+            if (!std::ranges::empty(get<0>(align)) && !std::ranges::empty(get<1>(align)))
+            {
+                ref_length = std::ranges::distance(get<1>(align));
 
-            for (auto chr : get<1>(align))
-                if (chr == gap{})
-                    ++off_end;
+                // compute possible distance from alignment end to sequence end
+                // which indicates soft clipping at the end.
+                // This should be replaced by a free count_gaps function for
+                // aligned sequences which is more efficient if possible.
+                assert(std::ranges::distance(seq) >= offset);
+                int32_t off_end{static_cast<int32_t>(std::ranges::distance(seq)) - offset};
 
-            off_end -= std::ranges::distance(get<1>(align));
+                for (auto chr : get<1>(align))
+                    if (chr == gap{})
+                        ++off_end;
 
-            std::vector<std::pair<char, size_t>> cigar = detail::get_cigar_vector(align, offset, off_end);
+                off_end -= ref_length;
```

Comment:
in case you change to unsigned, you need the assertion for this

Reply:
neither reflength nor off_end are unsigned